src/Enzo/enzo_EnzoMethodSinkMaker.cpp

```diff
@@ -0,0 +1,583 @@
+
+	// Check whether flow is converging
+	if (!flow_is_converging_(block, ix, iy, iz)) continue;
+
+	// If check_density_maximum_ is true, check if this cell's density is a local maximum
+	if (check_density_maximum_ && !density_is_local_maximum_(block,ix,iy,iz)) continue;
+
+	// If we are here, this cell satisfies all the conditions for forming a sink particle.
+
+	// Compute global cell index
+	const uint64_t global_cell_index = global_cell_index_start + block_cell_index;
+
+	// Append to the vectors
+	cells_forming_sinks.push_back(block_cell_index);
+	new_densities.push_back(density[block_cell_index] - density_change);
+
+	// So. now create a sink particle
+	n_sinks_formed++;
+
+	// ip_block is the index of the particle in the block
+	// ip_batch is the index if the particle in its batch
+	// ibatch is the index of the batch
+	int ip_block = particle.insert_particles(it, 1);
+	int ip_batch;
+	int ibatch;
+	particle.index(ip_block, &ibatch, &ip_batch);
+
+	// Set the mass of the sink particle to be sink_mass
+	pmass = (enzo_float *) particle.attribute_array(it, ia_m, ibatch);
+	pmass[ip_batch * dm] = sink_mass;
+
+	// Get 3 seeds for the random number generator using the global cell index and
+	// offset_seed_shift_
+	uint64_t x_seed = offset_seed_shift_ + 3 * global_cell_index;
+	uint64_t y_seed = offset_seed_shift_ + 3 * global_cell_index + 1;
+	uint64_t z_seed = offset_seed_shift_ + 3 * global_cell_index + 2;
+
+	// x_offset is in range [-hx*max_offset_cell_fraction_,hx*max_offset_cell_fraction_]
+	generator.seed(x_seed);
+	const double x_offset =
+	  hx * max_offset_cell_fraction_ * (2.0 * distribution(generator) - 1.0);
+	// y_offset is in range [-hy*max_offset_cell_fraction_,hy*max_offset_cell_fraction_]
+	generator.seed(y_seed);
+	const double y_offset =
+	  hy * max_offset_cell_fraction_ * (2.0 * distribution(generator) - 1.0);
+	// z_offset is in range [-hz*max_offset_cell_fraction_,hz*max_offset_cell_fraction_]
+	generator.seed(z_seed);
+	const double z_offset =
+	  hz * max_offset_cell_fraction_ * (2.0 * distribution(generator) - 1.0);
+
+	// Set particle position to be at centre of cell plus the offset
+	px = (enzo_float *) particle.attribute_array(it, ia_x, ibatch);
+	py = (enzo_float *) particle.attribute_array(it, ia_y, ibatch);
+	pz = (enzo_float *) particle.attribute_array(it, ia_z, ibatch);
+
+	px[ip_batch * dp] = xm + (ix - gx + 0.5) * hz + x_offset;
+	py[ip_batch * dp] = ym + (iy - gy + 0.5) * hy + y_offset;
+	pz[ip_batch * dp] = zm + (iz - gz + 0.5) * hz + z_offset;
+
+	// Set particle velocity equal to gas velocity in cell
+	pvx = (enzo_float *) particle.attribute_array(it, ia_vx, ibatch);
+	pvy = (enzo_float *) particle.attribute_array(it, ia_vy, ibatch);
+	pvz = (enzo_float *) particle.attribute_array(it, ia_vz, ibatch);
+	pvx[ip_batch * dv] = vx_gas[block_cell_index];
+	pvy[ip_batch * dv] = vy_gas[block_cell_index];
+	pvz[ip_batch * dv] = vz_gas[block_cell_index];
+
+	// Set creation time equal to current time
```

Comment:
Ah so this leads to a related question. When particles merge what is the creation time reset to?

Reply:
It's set to the minimum (i.e., the earliest).

Relevant code is on line 286 of `src/Enzo/enzo_EnzoMethodMergeSinks.cpp`:

```
 pcreation1 = std::min(pcreation1, pcreation2);
```